replica/table.cc

```diff
@@ -2103,9 +2106,29 @@ future<> tablet_storage_group_manager::update_effective_replication_map(const lo
     if (new_tablet_count > old_tablet_count) {
         tlogger.info0("Detected tablet split for table {}.{}, increasing from {} to {} tablets",
                         schema()->ks_name(), schema()->cf_name(), old_tablet_count, new_tablet_count);
-        return handle_tablet_split_completion(*old_tablet_map, *new_tablet_map);
+        co_await handle_tablet_split_completion(*old_tablet_map, *new_tablet_map);
+        co_return;
     }
-    return make_ready_future();
+
+    // Allocate storage group if tablet is migrating in.
+    auto holder = co_await lock().hold_write_lock();
+    auto this_replica = locator::tablet_replica{
+        .host = erm.get_token_metadata().get_my_id(),
+        .shard = this_shard_id()
+    };
+    auto tablet_migrates_in = [this_replica] (locator::tablet_transition_info& transition_info) {
+        return transition_info.stage == locator::tablet_transition_stage::allow_write_both_read_old && transition_info.pending_replica == this_replica;
+    };
+    for (auto& transition : new_tablet_map->transitions()) {
+        auto tid = transition.first;
+        auto transition_info = transition.second;
+        if (!_storage_groups.contains(tid.value()) && tablet_migrates_in(transition_info)) {
+            auto range = new_tablet_map->get_token_range(tid);
+            auto cg = std::make_unique<compaction_group>(_t, tid.value(), std::move(range));
+            _storage_groups[tid.value()] = std::make_unique<storage_group>(std::move(cg), &_compaction_groups);
+        }
+    }
```

Comment:
There's no preemption point between initialization of `holder` and here, so the why the lock? To wait on readers? This stuff is tricky and needs documentation.

Reply:
I think that to be on the safe side, and taking into account @raphaelsc's observation in https://github.com/scylladb/scylladb/pull/17664/commits/65e8ca669970e1d183d6a1f65bb93fa4164efee7#r1553599671
we can acquire the exclusive lock for the whole of `update_effective_replication_map`, both around `handle_tablet_split_completion` and the allocation/deallocation of storage_groups, based on the delta between the current tablet map and the updated one.
In addition, let's implement that in a separate function rather than in the body of this one - which should hold only the high level logic and then fan out to lower level functions, depending on what the update requires.